llmfoundry/callbacks/hf_checkpointer.py

```diff
@@ -225,7 +233,18 @@ def _save_checkpoint(self, state: State, logger: Logger):
                         )
 
                 elapsed_duration = state.get_elapsed_duration()
-                if self.mlflow_registered_model_name is not None and elapsed_duration is not None and elapsed_duration >= 1.0:
+
+                # If the save interval is specified as 1dur, and the max duration is in epoch units
+                # we need a special case to identify we are on the last batch and should write the mlflow checkpoint
+                is_last_batch = False
+                assert state.max_duration is not None  # for pyright
+                if self.save_interval.unit == TimeUnit.DURATION and self.save_interval.value == 1 and state.max_duration.unit == TimeUnit.EPOCH:
+                    assert state.dataloader_len is not None  # for pyright
+                    is_last_batch = int(state.timestamp.batch) % math.ceil(
+                        state.max_duration.value * state.dataloader_len) == 0
+                if self.mlflow_registered_model_name is not None and (
+                    (elapsed_duration is not None and
+                     elapsed_duration >= 1.0) or is_last_batch):
```

Comment:
nit: can we use a helper function to define is_last_batch? I think it'll be more clear that way. The conditional here can be updated to`if self.mlflow_registered_model_name is not None and _is_last_batch(state, self.save_interval)` 

```
def _is_last_batch(state, save_interval):
     elapsed_duration = state.get_elapsed_duration()
     if elapsed_duration is not None and elapsed_duration >= 1.0:
         return True
     assert state.max_duration is not None
     if self.save_interval.unit == TimeUnit.DURATION and self.save_interval.value == 1 and state.max_duration.unit == TimeUnit.EPOCH:
                    assert state.dataloader_len is not None  # for pyright
                     return int(state.timestamp.batch) % math.ceil(
                        state.max_duration.value * state.dataloader_len) == 0
      return False
```